File: LeetCodeCPlusPlus/py2.cpp
#include<iostream>
#include<vector>
using namespace std;
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
 };
 
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* head = new ListNode(0);
        ListNode* temp = head;
        int carry = 0;
        while(l1 != NULL || l2 != NULL) {
            int oper1 = (l1 == NULL) ? 0 : l1->val;
            int oper2 = (l2 == NULL) ? 0 : l2->val;
            int operSum = oper1 + oper2 + carry;
            carry = operSum / 10;
            int add = operSum % 10;
            temp->next = new ListNode(add);
            temp = temp->next;
            if(l1 != NULL) l1 = l1->next;
            if(l2 != NULL) l2 = l2->next;
        }
        if(carry>0) {
            temp->next = new ListNode(carry);
        }
        return head->next;
    }

    ListNode* construct(vector<int> a) {
        ListNode* head = new ListNode(0);
        ListNode* temp = head;
        for (int i=0;i<a.size();i++) {
            temp->next = new ListNode(a[i]);
            temp = temp->next;
        }
        return head->next;
    }
};

int main(){
    Solution solution = Solution();
    vector<int> a = {2, 7, 1, 5};
    vector<int> b = {3, 3, 2, 4};
    ListNode* l1 = solution.construct(a);
    ListNode* l2 = solution.construct(b);
    ListNode* ans = solution.addTwoNumbers(l1, l2);
    while (ans != NULL) {
        cout << ans->val << " ";
        ans = ans->next;
    }
    return 0;

}

File: LeetCodePython/py2.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def mySum(self, x, y):
        return (x+y)%10, (x+y)//10
    
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        
        if not l1 and not l2:
            return None
        
        if not l1:
            return l2
        elif not l2:
            return l1
        
        result = ListNode(0)
        pre = result
        plus = 0
        while(l1 and l2):
            total, plus = self.mySum(l1.val+plus, l2.val) 
            q = ListNode(total)
            pre.next = q
            pre = q
            l1 = l1.next
            l2 = l2.next
        
        while l1:
            total,plus = self.mySum(l1.val, plus)
            q = ListNode(total)
            pre.next = q
            pre = q
            l1 = l1.next
        
        while l2:
            total,plus = self.mySum(l2.val, plus)
            q = ListNode(total)
            pre.next = q
            pre = q
            l2 = l2.next
        
        if plus:
            q = ListNode(plus)
            pre.next = q
            pre = q
            
        pre.next = None
            
        return result.next

def construct(data:list):
    head = ListNode(0)
    pre = head
    for item in data:
        temp = ListNode(item)
        pre.next = temp
        pre = temp
    return head.next

if __name__ == '__main__':
    import time
    start = time.process_time()
    a = [2,4,3]
    b = [5,6,4]
    l1 = construct(a)
    l2 = construct(b)
    ans = Solution().addTwoNumbers(l1, l2)
    while ans:
        print(ans.val, end=" ")
        ans = ans.next
    end = time.process_time()
    print("run time:",str(end-start))
